### PR TITLE
feat: Add trivy scanning back to VHD build

### PR DESCRIPTION
**Pull request by AlisonB319**
**What type of PR is this?**
Trivy is needed back in our pipeline builds. This uses a new method of trivy scanning by creating a vm based of the just-built vhd, running the trivy scans, uploading the generated files from the vm to blob storage and then downloading the files from blob storage back into the pipeline

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Requirements**:
- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:
```
none
```
example run with new artifact layout
https://msazure.visualstudio.com/CloudNativeCompute/_build/results?buildId=95012546&view=artifacts&pathAsName=false&type=publishedArtifacts

[trivy-images-table.txt](https://github.com/user-attachments/files/15506841/trivy-images-table.11.txt)
[trivy-report.json](https://github.com/user-attachments/files/15506842/trivy-report.5.json)

Example of trivy-images-table.txt
```
This contains the list of images with high and critical level CVEs (if present), that are present in the node. 
Note: images without CVEs are also listed

mcr.microsoft.com/aks/aks-node-ca-watcher:master.221011.1 (debian 11.5)
=======================================================================
Total: 0 (HIGH: 0, CRITICAL: 0)


mcr.microsoft.com/aks/ip-masq-agent-v2:v0.1.10 (debian 12.4)
============================================================
Total: 2 (HIGH: 2, CRITICAL: 0)

┌─────────┬────────────────┬──────────┬───────────────────┬────────────────┬────────────────────────────────────────────────────────┐
│ Library │ Vulnerability  │ Severity │ Installed Version │ Fixed Version  │                         Title                          │
├─────────┼────────────────┼──────────┼───────────────────┼────────────────┼────────────────────────────────────────────────────────┤
│ libc6   │ CVE-2024-2961  │ HIGH     │ 2.36-9+deb12u4    │ 2.36-9+deb12u6 │ glibc: Out of bounds write in iconv may lead to remote │
│         │                │          │                   │                │ code...                                                │
│         │                │          │                   │                │ https://avd.aquasec.com/nvd/cve-2024-2961              │
│         ├────────────────┤          │                   ├────────────────┼────────────────────────────────────────────────────────┤
│         │ CVE-2024-33599 │          │                   │ 2.36-9+deb12u7 │ glibc: stack-based buffer overflow in netgroup cache   │
│         │                │          │                   │                │ https://avd.aquasec.com/nvd/cve-2024-33599             │
└─────────┴────────────────┴──────────┴───────────────────┴────────────────┴────────────────────────────────────────────────────────┘
```
example of trivy report 

```
{
  "SchemaVersion": 2,
  "ArtifactName": "vhd-scanning-vm",
  "ArtifactType": "filesystem",
  "Metadata": {
    "OS": {
      "Family": "cbl-mariner",
      "Name": "2.0.20240425"
    },
    "ImageConfig": {
      "architecture": "",
      "created": "0001-01-01T00:00:00Z",
      "os": "",
      "rootfs": {
        "type": "",
        "diff_ids": null
      },
      "config": {}
    }
  },
  "Results": [
```

### CODE REVIEW

File: vhdbuilder/packer/vhd-scanning.sh [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    VM_OPTIONS="--size Standard_D2pds_v5"
+elif [[ "${FEATURE_FLAGS,,}" == "kata" ]]; then
+    VM_OPTIONS="--size Standard_D4ds_v5"
+fi
+
+if [[ "${OS_TYPE}" == "Linux" && "${ENABLE_TRUSTED_LAUNCH}" == "True" ]]; then
+    VM_OPTIONS+="--security-type TrustedLaunch --enable-secure-boot true --enable-vtpm true"
+fi
+
+#fix identity string
```
Comment (djsly) [inline vhdbuilder/packer/vhd-scanning.sh:42]: //FIXME. ?   we should come up with a fixed set of acronyms which works with VSCODE

File: vhdbuilder/packer/vhd-scanning.sh [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+az vm create --resource-group $RESOURCE_GROUP_NAME \
+    --name $VM_NAME \
+    --image $VHD_IMAGE \
+    --admin-username $TEST_VM_ADMIN_USERNAME \
+    --admin-password $TEST_VM_ADMIN_PASSWORD \
+    --os-disk-size-gb 50 \
+    --assign-identity "/subscriptions/8ecadfc9-d1a3-4ea4-b844-0d9f87e4d7c8/resourceGroups/aksvhdtestbuildrg/providers/Microsoft.ManagedIdentity/userAssignedIdentities/vhd-scanning-UAMI" \
```
Comment (djsly) [inline vhdbuilder/packer/vhd-scanning.sh:49]: are you planning to turn this into a var before merging ?

File: vhdbuilder/packer/vhd-scanning.sh [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+FULL_PATH=$(realpath $0)
+CDIR=$(dirname $FULL_PATH)
+SCRIPT_PATH="$CDIR/$TRIVY_SCRIPT_PATH"
+az vm run-command invoke \
+    --command-id RunShellScript \
+    --name $VM_NAME \
+    --resource-group $RESOURCE_GROUP_NAME \
+    --scripts @$SCRIPT_PATH
+
+if [[ "$OS_SKU" = "Ubuntu" && "$SIG_IMAGE_NAME" == *2004* ]]; then
```
Comment (djsly) [inline vhdbuilder/packer/vhd-scanning.sh:61]: could we look at turning all those `az vm run-command invoke` into a single shell scripts ?

```
   az vm run-command invoke \
         --command-id RunShellScript \
         --name $VM_NAME \
         --resource-group $RESOURCE_GROUP_NAME \
         --scripts @$UPLOAD_SCANNING_RESULTS_SCRIPT_PATH
```

**Review by djsly — CHANGES_REQUESTED** [review]
Could you provide examples of the two files being upload in the PR description ? thanks !

File: vhdbuilder/packer/vhd-scanning.sh (hunk 1 above)
Comment (AlisonB319) [reply vhdbuilder/packer/vhd-scanning.sh:42]: This is just a small reminder for myself - I'll have it cleaned up after this PR goes out of draft

File: vhdbuilder/packer/vhd-scanning.sh [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+az vm create --resource-group $RESOURCE_GROUP_NAME \
+    --name $VM_NAME \
+    --image $VHD_IMAGE \
+    --admin-username $TEST_VM_ADMIN_USERNAME \
+    --admin-password $TEST_VM_ADMIN_PASSWORD \
+    --os-disk-size-gb 50 \
+    --assign-identity "/subscriptions/8ecadfc9-d1a3-4ea4-b844-0d9f87e4d7c8/resourceGroups/aksvhdtestbuildrg/providers/Microsoft.ManagedIdentity/userAssignedIdentities/vhd-scanning-identity" \
+    ${VM_OPTIONS}
+
+if [[ "${ENABLE_TRUSTED_LAUNCH}" == "True" ]]; then
```
Comment (AlisonB319) [inline vhdbuilder/packer/vhd-scanning.sh:54]: Here is where I attempted it

**Review by djsly — APPROVED** [review]

**Review by djsly — APPROVED** [review]

**Pull request merged**